### ISSUE TITLE
Scraper request: bergamot.app

### ISSUE THREAD

**Issue by josefhelie** [hhursev/recipe-scrapers#986]
I'm currently using the free app Bergamot (which is closed source) to store my recipes, but I'd like to move to Mealie. I've encountered an error message that says, 'recipe_scrapers was unable to scrape this URL.' Is it possible to get a scraper, please? 😇
Thanks for your help.
A link to a shared recipe: https://dashboard.bergamot.app/shared/T8IJLjbtHdh2pj

**Comment by josefhelie** [comment]
i'm sorry I shared a recipe that don't reflect all the requested fields. Here is a better example: https://dashboard.bergamot.app/shared/mIB4jYQtZU1A97
Is it better?

**Comment by jayaddison** [comment]
Yep, that initially looks good to me @josefhelie - it's difficult to say for certain without coding it up, but it seems to have most/all of the information we'd need.  Thanks!

**Comment by josefhelie** [comment]
Thanks a lot @jayaddison :)

**Comment by jayaddison** [comment]
Hi @josefhelie - apologies for my delayed reply.  No further updates on this at the moment I'm afraid.  Do you have any interest in learning some Python coding?

**Comment by mlduff** [comment]
@jayaddison I took a look, looks like it is fairly easy to call the API endpoint, which can be derived from the URL of the recipe.
For https://dashboard.bergamot.app/shared/mIB4jYQtZU1A97 the associated API endpoint is https://api.bergamot.app/recipes/shared?r=mIB4jYQtZU1A97.

I'm not sure how the library normally supports the case of recipes being loaded via an API call after the original page load, but I'd be happy to take this task on if you like?

**Comment by jayaddison** [comment]
Thanks @mlduff!

> I'm not sure how the library normally supports the case of recipes being loaded via an API call after the original page load - I can see a few examples (`goustojson.py`, `monsieurcuisine.py`) that seem to do this - I would be happy to tackle this if you are happy to me to do so?

About the handling of APIs: yep, well discovered - we do have a few scrapers that retrieve data using APIs at the moment.  A potential design/architecture problem with that is that it (currently) tightly-couples the scraper to an HTTP client - namely [`requests`](https://github.com/psf/requests) at the moment; nearly a de-facto client for Python, but even so, it may not be ideal to depend entirely on it.

Meanwhile we have a [`v15` development branch](https://github.com/hhursev/recipe-scrapers/tree/v15) that can optionally use `requests`, but that otherwise requires callers to retrieve the HTML and pass it to the scraper themselves.  Marginally less convenient, but allowing callers to use whatever HTTP client(s) they prefer (anything from built-in `urlopen`, low-level `urllib3`, `requests`, `httpx`, etc).

A long explanation, but the short answer is: yep, please go ahead, but be aware that this would currently only be supported in the v14 / mainline branch.

**Comment by jayaddison** [comment]
@mlduff also a design / implementation question for your consideration: those recipes sometimes contain a link to the original source of the recipe.  Should we return that as the [canonical URL](https://en.wikipedia.org/wiki/Canonical_link_element) for recipes when possible?

**Comment by mlduff** [comment]
> Meanwhile we have a [`v15` development branch](https://github.com/hhursev/recipe-scrapers/tree/v15) that can optionally use `requests`, but that otherwise requires callers to retrieve the HTML and pass it to the scraper themselves. Marginally less convenient, but allowing callers to use whatever HTTP client(s) they prefer (anything from built-in `urlopen`, low-level `urllib3`, `requests`, `httpx`, etc).

Is your preference for me to develop this in the v15 branch? If I implement in v14 (which seems easier), will it then need rewriting at some point (are the other ones like the example I found going to also need similar rewriting?)?

**Comment by mlduff** [comment]
> @mlduff also a design / implementation question for your consideration: those recipes sometimes contain a link to the original source of the recipe. Should we return that as the [canonical URL](https://en.wikipedia.org/wiki/Canonical_link_element) for recipes when possible?

Good point, will try to do that.

**Comment by jayaddison** [comment]
> > Meanwhile we have a [`v15` development branch](https://github.com/hhursev/recipe-scrapers/tree/v15) that can optionally use `requests`, but that otherwise requires callers to retrieve the HTML and pass it to the scraper themselves. Marginally less convenient, but allowing callers to use whatever HTTP client(s) they prefer (anything from built-in `urlopen`, low-level `urllib3`, `requests`, `httpx`, etc).
> 
> @jayaddison is your preference for me to develop this in the v15 branch? If I implement in v14 (which seems easier), will it then need rewriting at some point (are the other ones like the example I found going to also need similar rewriting?)?

I'd recommend implementing it for `v14`, yep.

**Comment by josefhelie** [comment]
> Hi @josefhelie - apologies for my delayed reply. No further updates on this at the moment I'm afraid. Do you have any interest in learning some Python coding?
Thanks @jayaddison, but i don't have enough free time to do that, even if I would like to!! 😢
Thanks @mlduff too :)

**Comment by mlduff** [comment]
@jayaddison I noticed that the tests for the two scrapers I mentioned above are located under the legacy section - do I add my tests under there as well?

**Comment by mlduff** [comment]
@josefhelie are you able to provide a couple more recipe URLs please so I can test?

**Comment by jayaddison** [comment]
> @jayaddison I noticed that the tests for the two scrapers I mentioned above are located under the legacy section - do I add my tests under there as well?

@mlduff yep, that's the correct place for those; thanks for checking :+1:  You should be able to configure the `expected_requests` property in the tests to return example results for both the initial HTML `HTTP GET` response, and also the subsequent (probably also `HTTP GET`) API request.

**Comment by jayaddison** [comment]
@josefhelie have you found any pages shared on Bergamot where the original author is credited?  I've seen a few pages that have the domain name of the source URL.. I'm wondering whether there are any that list names/usernames.

**Comment by josefhelie** [comment]
@jayaddison I'm not sure I have. Would it help you if you provide me a recipe I could import into Bergamot and then give you the link towards the imported recipe?

**Comment by mlduff** [comment]
@josefhelie Here is one that has an author https://www.bestrecipes.com.au/recipes/peanut-butter-cookies-recipe/fowk6kuy

**Comment by josefhelie** [comment]
I imported it in my Bergamot, here it is: https://dashboard.bergamot.app/shared/REbGkQaNoVJ5kM
